qiskit/quantum_info/synthesis/qsd.py

```diff
@@ -0,0 +1,191 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Quantum Shannon Decomposition.
+
+Method is described in arXiv:quant-ph/0406176.
+"""
+import scipy
+import numpy as np
+from qiskit.circuit import QuantumCircuit, QuantumRegister
+from qiskit.quantum_info.synthesis import two_qubit_decompose, one_qubit_decompose
+from qiskit.circuit.library.standard_gates import CXGate
+from qiskit.quantum_info.operators.predicates import is_hermitian_matrix
+from qiskit.extensions.quantum_initializer.uc_pauli_rot import UCPauliRotGate, _EPS
+
+
+def qs_decomposition(mat, opt_a1=True, decomposer_1q=None, decomposer_2q=None):
+    """
+    Decomposes unitary matrix into one and two qubit gates using Quantum Shannon Decomposition.
+
+       ┌───┐               ┌───┐     ┌───┐     ┌───┐
+      ─┤   ├─       ───────┤ Rz├─────┤ Ry├─────┤ Rz├─────
+       │   │    ≃     ┌───┐└─┬─┘┌───┐└─┬─┘┌───┐└─┬─┘┌───┐
+     /─┤   ├─       /─┤   ├──□──┤   ├──□──┤   ├──□──┤   ├
+       └───┘          └───┘     └───┘     └───┘     └───┘
+
+    The number of CX gates generated with the decomposition without optimizations is,
+
+    .. math::
+
+        \frac{9}{16} 4^n - frac{3}{2} 2^n
+
+    If opt_a1=True, the CX count is further reduced by,
+
+    .. math::
+
+        \frac{1}{3} 4^{n - 2} - 1
+
+    This decomposition is described in arXiv:quant-ph/0406176.
+
+    Arguments:
+       mat (ndarray): unitary matrix to decompose
+       opt_a1 (bool): whether to try optimization A.1 from Shende. This should elliminate 1 cnot
+          per call. If True CZ gates are left in the output. If desired these can be further decomposed
```

Comment:
```suggestion
       opt_a1 (bool): whether to try optimization A.1 from Shende. This should eliminate 1 cnot
          per call. If True CZ gates are left in the output. If desired these can be further decomposed
```

Reply:
also what do you mean "eliminate 1 cnot per call"? should it be called multiple times?